tsl/src/compression/compression.c

```diff
@@ -2042,3 +2047,226 @@ update_compressed_chunk_relstats(Oid uncompressed_relid, Oid compressed_relid)
 		CommandCounterIncrement();
 	}
 }
+
+/* Returns a Bitmap with all columns part of unique constraints */
+static Bitmapset *
+relation_unique_columns(Relation rel)
+{
+	/*
+	 * We are using the list of indexes to determine columns
+	 * part of unique constraints here because the index list
+	 * is cached. It might be more correct to use the constraints
+	 * of the relation but getting those requires a sequential
+	 * scan on the catalog.
+	 */
+	List *indexoidlist = RelationGetIndexList(rel);
+	ListCell *lc;
+
+	/* Bitmapset of unique columns in the uncompressed chunk */
+	Bitmapset *unique_columns = NULL;
+
+	foreach (lc, indexoidlist)
+	{
+		Oid indexoid = lfirst_oid(lc);
+		HeapTuple index_tuple;
+		Form_pg_index index;
+
+		index_tuple = SearchSysCache1(INDEXRELID, ObjectIdGetDatum(indexoid));
+		if (!HeapTupleIsValid(index_tuple))
+			elog(ERROR,
+				 "cache lookup failed for index %u in \"%s\" ",
+				 indexoid,
+				 RelationGetRelationName(rel));
+		index = (Form_pg_index) GETSTRUCT(index_tuple);
+
+		if (index->indisunique)
+		{
+			for (int i = 0; i < index->indnkeyatts; i++)
+			{
+				unique_columns = bms_add_member(unique_columns, index->indkey.values[i]);
+			}
+		}
+
+		ReleaseSysCache(index_tuple);
+	}
+
+	list_free(indexoidlist);
+
+	return unique_columns;
+}
+
+static ScanKeyData *
+build_scankeys(int32 hypertable_id, RowDecompressor decompressor, Bitmapset *unique_columns,
+			   Bitmapset **null_columns, TupleTableSlot *slot, int *num_scankeys)
+{
+	int index = 0;
+	ScanKeyData *scankeys = NULL;
+
+	if (!bms_is_empty(unique_columns))
+	{
+		scankeys = palloc0(bms_num_members(unique_columns) * 2 * sizeof(ScanKeyData));
+		int attno = -1;
+		while ((attno = bms_next_member(unique_columns, attno)) > 0)
+		{
+			char *attname = get_attname(decompressor.out_rel->rd_id, attno, false);
+			FormData_hypertable_compression *fd =
+				ts_hypertable_compression_get_by_pkey(hypertable_id, attname);
+
+			/*
+			 * There are 3 possible scenarios we have to consider
+			 * when dealing with columns which are part of unique
+			 * constraints.
+			 *
+			 * 1. Column is segmentby-Column
+			 * In this case we can add a single ScanKey with an
+			 * equality check for the value.
+			 * 2. Column is orderby-Column
+			 * In this we can add 2 ScanKeys with range constraints
+			 * utilizing batch metadata.
+			 * 3. Column is neither segmentby nor orderby
+			 * In this case we cannot utilize this column for
+			 * batch filtering as the values are compressed and
+			 * we have no metadata.
+			 */
+
+			if (COMPRESSIONCOL_IS_SEGMENT_BY(fd))
+			{
+				bool isnull;
+				Datum value = slot_getattr(slot, attno, &isnull);
+
+				TypeCacheEntry *tce =
+					lookup_type_cache(decompressor.in_desc->attrs[attno - 1].atttypid,
```

Comment:
`attno` is taken from the uncompressed chunk. is it guaranteed to be the same for the compressed? or should we be using the attname again to get the attno for the compressed chunk?

Reply:
good find, we do indeed need to do proper mapping here

Reply:
Any reason not to convert `attno - 1` to `AttrNumberGetAttrOffset(attno)`?

Reply:
done